Assertion error should be raised when you have non-string response header

``` python
from __future__ import unicode_literals
import unittest
from wsgiref.simple_server import make_server

from webtest import TestApp
from pyramid.config import Configurator
from pyramid.response import Response


def hello_world(request):
    return Response("{'foo': 'bar'}", content_type='application/json')

def main():
    config = Configurator()
    config.add_route('hello', '/')
    config.add_view(hello_world, route_name='hello')
    app = config.make_wsgi_app()
    return app


class TestAppCase(unittest.TestCase):

    def test_unicode_header_should_raise_error(self):
        app = main()
        test_app = TestApp(app)
        test_app.get('/', status=500)


if __name__ == '__main__':
    app = main()
    server = make_server('0.0.0.0', 8080, app)
    server.serve_forever()
```

here is a demo, when you run the app like this

```
python demo.py
```

Then visit that server

```
curl localhost:8080
```

then you get a traceback like this

```
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/.env/lib/python2.7/site-packages/pyramid/router.py", line 243, in __call__
    return response(request.environ, start_response)
  File "/.env/lib/python2.7/site-packages/webob/response.py", line 1021, in __call__
    start_response(self.status, headerlist)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/wsgiref/handlers.py", line 180, in start_response
    assert type(val) is StringType,"Header values must be strings"
AssertionError: Header values must be strings
```

since the `content_type` we give there

```
content_type='application/json'
```

is an unicode. However, when you run the tests

```
nosetests -sv demo.py
```

you got

```
AppError: Bad response: 200 OK (not 500)
```

It doesn't raise an assertion. I think this is pretty bad, I always need to run my application in real environment so that I know I have unicode passed to the response. Instead of deferring error spotting to production environment, I suggest we should check the response value, see is there unicode in header.

--- a/tests/test_forms.py
+++ b/tests/test_forms.py
@@ -321,6 +321,8 @@ def select_app(environ, start_response):
     headers = [
         ('Content-Type', 'text/html; charset=utf-8'),
         ('Content-Length', str(len(body)))]
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(status, headers)
     return [body]
 
@@ -373,6 +375,8 @@ def select_app_without_values(environ, start_response):
     headers = [
         ('Content-Type', 'text/html; charset=utf-8'),
         ('Content-Length', str(len(body)))]
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(status, headers)
     return [body]
 
@@ -425,6 +429,8 @@ def select_app_without_default(environ, start_response):
     headers = [
         ('Content-Type', 'text/html; charset=utf-8'),
         ('Content-Length', str(len(body)))]
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(status, headers)
     return [body]
 
@@ -476,6 +482,8 @@ def select_app_unicode(environ, start_response):
     headers = [
         ('Content-Type', 'text/html; charset=utf-8'),
         ('Content-Length', str(len(body)))]
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(status, headers)
     if not isinstance(body, binary_type):
         raise AssertionError('Body is not %s' % binary_type)
@@ -770,6 +778,8 @@ class SingleUploadFileApp(object):
         headers = [
             ('Content-Type', 'text/html; charset=utf-8'),
             ('Content-Length', str(len(body)))]
+        # PEP 3333 requires native strings:
+        headers = [(str(k), str(v)) for k, v in headers]
         start_response(status, headers)
         assert(isinstance(body, binary_type))
         return [body]

--- a/tests/test_lint.py
+++ b/tests/test_lint.py
@@ -143,13 +143,33 @@ class TestCheckContentType(unittest.TestCase):
 
 class TestCheckHeaders(unittest.TestCase):
 
-    @unittest.skipIf(not PY3, 'Useless in Python2')
-    def test_header_unicode_value(self):
-        self.assertRaises(AssertionError, check_headers, [('X-Price', '100€')])
-
-    @unittest.skipIf(not PY3, 'Useless in Python2')
+    @unittest.skipIf(PY3, 'unicode is str in Python3')
     def test_header_unicode_name(self):
-        self.assertRaises(AssertionError, check_headers, [('X-€', 'foo')])
+        headers = [(u'X-Price', str('100'))]
+        self.assertRaises(AssertionError, check_headers, headers)
+
+    @unittest.skipIf(PY3, 'unicode is str in Python3')
+    def test_header_unicode_value(self):
+        headers = [(str('X-Price'), u'100')]
+        self.assertRaises(AssertionError, check_headers, headers)
+
+    @unittest.skipIf(not PY3, 'bytes is str in Python2')
+    def test_header_bytes_name(self):
+        headers = [(b'X-Price', '100')]
+        self.assertRaises(AssertionError, check_headers, headers)
+
+    @unittest.skipIf(not PY3, 'bytes is str in Python2')
+    def test_header_bytes_value(self):
+        headers = [('X-Price', b'100')]
+        self.assertRaises(AssertionError, check_headers, headers)
+
+    def test_header_non_latin1_value(self):
+        headers = [(str('X-Price'), '100€')]
+        self.assertRaises(AssertionError, check_headers, headers)
+
+    def test_header_non_latin1_name(self):
+        headers = [('X-€', str('foo'))]
+        self.assertRaises(AssertionError, check_headers, headers)
 
 
 class TestCheckEnviron(unittest.TestCase):

--- a/tests/test_lint.py
+++ b/tests/test_lint.py
@@ -114,6 +114,8 @@ class TestMiddleware2(unittest.TestCase):
             headers = [
                 ('Content-Type', 'text/plain; charset=utf-8'),
                 ('Content-Length', str(len(body)))]
+            # PEP 3333 requires native strings:
+            headers = [(str(k), str(v)) for k, v in headers]
             start_response(to_bytes('200 OK'), headers, ('stuff',))
             return [body]
 

--- a/tests/test_response.py
+++ b/tests/test_response.py
@@ -115,7 +115,8 @@ def links_app(environ, start_response):
     ]
     if req.path_info in utf8_paths:
         headers[0] = ('Content-Type', str('text/html; charset=utf-8'))
-
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(str(status), headers)
     return [body]
 
@@ -127,7 +128,8 @@ def gzipped_app(environ, start_response):
         ('Content-Type', str('text/html')),
         ('Content-Encoding', str('gzip')),
     ]
-
+    # PEP 3333 requires native strings:
+    headers = [(str(k), str(v)) for k, v in headers]
     start_response(str(status), headers)
     return encoded_body
 
@@ -423,6 +425,8 @@ class TestFollow(unittest.TestCase):
                 remaining_redirects[0] -= 1
 
             headers.append(('Content-Length', str(len(body))))
+            # PEP 3333 requires native strings:
+            headers = [(str(k), str(v)) for k, v in headers]
             start_response(str(status), headers)
             return [body]
 

--- a/webtest/lint.py
+++ b/webtest/lint.py
@@ -59,7 +59,8 @@ Some of the things this checks:
 * That the headers is a list (not a subclass, not another kind of
   sequence).
 
-* That the items of the headers are tuples of strings.
+* That the items of the headers are tuples of 'native' strings (i.e.
+  bytestrings in Python2, and unicode strings in Python3).
 
 * That there is no 'status' header (that is used in CGI, but not in
   WSGI).
@@ -412,12 +413,12 @@ def check_status(status):
         "digits and a space (4th characters is not a space here)") % status
 
 
-def _assert_latin1_py3(string, message):
-    if PY3 and type(string) is str:
-        try:
-            string.encode('latin1')
-        except UnicodeEncodeError:
-            raise AssertionError(message)
+def _assert_latin1_str(string, message):
+    assert type(string) is str, message
+    try:
+        string.encode('latin1')
+    except UnicodeEncodeError:
+        raise AssertionError(message)
 
 
 def check_headers(headers):
@@ -430,30 +431,30 @@ def check_headers(headers):
             % (item, type(item)))
         assert len(item) == 2
         name, value = item
-        _assert_latin1_py3(
+        _assert_latin1_str(
             name,
-            "Headers values must be latin1 string or bytes."
-            "%r is not a valid latin1 string" % (value,)
+            "Header names must be latin1 string "
+            "(not Py2 unicode or Py3 bytes type). "
+            "%r is not a valid latin1 string" % (name,)
         )
-        str_name = to_string(name)
-        assert str_name.lower() != 'status', (
+        assert name.lower() != 'status', (
             "The Status header cannot be used; it conflicts with CGI "
             "script, and HTTP status is not given through headers "
             "(value: %r)." % value)
-        assert '\n' not in str_name and ':' not in str_name, (
+        assert '\n' not in name and ':' not in name, (
             "Header names may not contain ':' or '\\n': %r" % name)
-        assert header_re.search(str_name), "Bad header name: %r" % name
-        assert not str_name.endswith('-') and not str_name.endswith('_'), (
+        assert header_re.search(name), "Bad header name: %r" % name
+        assert not name.endswith('-') and not name.endswith('_'), (
             "Names may not end in '-' or '_': %r" % name)
-        _assert_latin1_py3(
+        _assert_latin1_str(
             value,
-            "Headers values must be latin1 string or bytes."
+            "Header values must be latin1 string "
+            "(not Py2 unicode or Py3 bytes type)."
             "%r is not a valid latin1 string" % (value,)
         )
-        str_value = to_string(value)
-        assert not bad_header_value_re.search(str_value), (
+        assert not bad_header_value_re.search(value), (
             "Bad header value: %r (bad char: %r)"
-            % (str_value, bad_header_value_re.search(str_value).group(0)))
+            % (str_value, bad_header_value_re.search(value).group(0)))
 
 
 def check_content_type(status, headers):
@@ -464,13 +465,11 @@ def check_content_type(status, headers):
     NO_MESSAGE_TYPE = (204, 304)
     length = None
     for name, value in headers:
-        str_name = to_string(name)
-        if str_name.lower() == 'content-length' and value.isdigit():
+        if name.lower() == 'content-length' and value.isdigit():
             length = int(value)
             break
     for name, value in headers:
-        str_name = to_string(name)
-        if str_name.lower() == 'content-type':
+        if name.lower() == 'content-type':
             if code not in NO_MESSAGE_TYPE:
                 return
             elif length == 0:

